tensorflow_model_optimization/python/examples/sparsity/keras/mnist/mnist_e2e_sparsity2x4.py

```diff
@@ -0,0 +1,150 @@
+epochs = 1
+
+PRUNABLE_2x4_LAYERS = (tf.keras.layers.Conv2D, tf.keras.layers.Dense)
+
+
+def check_model_sparsity_2x4(model):
+  for layer in model.layers:
+    if isinstance(layer, pruning_wrapper.PruneLowMagnitude) and\
+      isinstance(layer.layer, PRUNABLE_2x4_LAYERS):
+      for weight in layer.layer.get_prunable_weights():
+        if pruning_utils.check_if_applicable_sparsity_2x4(weight) and\
+          not pruning_utils.is_pruned_2x4(weight):
+            return False
+  return True
+
+
+def build_layerwise_model(input_shape, **pruning_params):
+  return tf.keras.Sequential([
+      prune.prune_low_magnitude(l.Conv2D(
+          32, 5, padding='same',
+          activation='relu',
+          input_shape=input_shape),
+          **pruning_params),
+      l.MaxPooling2D((2, 2), (2, 2), padding='same'),
+      prune.prune_low_magnitude(l.Conv2D(
+          64, 5, padding='same'),
+          **pruning_params),
+      l.BatchNormalization(),
+      l.ReLU(),
+      l.MaxPooling2D((2, 2), (2, 2), padding='same'),
+      l.Flatten(),
+      prune.prune_low_magnitude(
+          l.Dense(1024,
+          activation='relu'
+          ), **pruning_params),
+      l.Dropout(0.4),
+      l.Dense(num_classes, activation='softmax')
+  ])
+
+
+def train(model, x_train, y_train, x_test, y_test):
+  model.compile(
+      loss=tf.keras.losses.categorical_crossentropy,
+      optimizer='adam',
+      metrics=['accuracy'])
+  model.run_eagerly = True
+
+  # Print the model summary.
+  model.summary()
+
+  # Add a pruning step callback to peg the pruning step to the optimizer's
+  # step. Also add a callback to add pruning summaries to tensorboard
+  callbacks = [
+      pruning_callbacks.UpdatePruningStep(),
+      pruning_callbacks.PruningSummaries(log_dir='/tmp/logs')
+  ]
+
+  model.fit(
+      x_train,
+      y_train,
+      batch_size=batch_size,
+      epochs=epochs,
+      verbose=1,
+      callbacks=callbacks,
+      validation_data=(x_test, y_test))
+  score = model.evaluate(x_test, y_test, verbose=0)
+  print('Test loss:', score[0])
+  print('Test accuracy:', score[1])
+
+  # Check sparsity 2x4 type before stripping pruning
+  is_pruned_2x4 = check_model_sparsity_2x4(model)
+  print("Pass the check for sparsity 2x4: ", is_pruned_2x4)
+
+  model = prune.strip_pruning(model)
+  return model
+
+
+def main(unused_argv):
+  ##############################################################################
+  # Prepare training and testing data
+  ##############################################################################
+  (x_train, y_train), (
+      x_test,
+      y_test), input_shape = keras_test_utils.get_preprocessed_mnist_data()
+
+  ##############################################################################
+  # Train a model with sparsity 2x4.
+  ##############################################################################
+  pruning_params = {
+      'pruning_schedule':
+          ConstantSparsity(0.5, begin_step=0, frequency=100),
```

Comment:
Can you briefly explain the pruning algorithm with this implementation? If I understood correctly, the algorithm Nvidia proposed will only mask the weight once at the beginning of training. The mask is then fixed throughout the training process. Since we have the pruning schedule configured here, does that mean the mask could be updated later?

Reply:
we are doing the same essentially here as Nvidia. we compute mask at the beginning and then it stays the same during the training. i agree that it might be confusing with this scheduler. our intention was to try to increase sparsity gradually, but this work is still under development.